xdsl/traits.py

```diff
@@ -446,3 +442,59 @@ def verify(self, op: Operation) -> None:
     @abc.abstractmethod
     def get_canonicalization_patterns(cls) -> tuple[RewritePattern, ...]:
         raise NotImplementedError()
+
+
+class MemoryEffect(OpTrait):
+    """
+    A trait that enables operations to expose their side-effects or absence thereof.
+
+    NB: The MLIR implementation further allows to describe what *kind* of side-effects
+    an operation has, e.g., read-only, or allocation.
+    This one is a stripped down version for now, just saying if there are any
+    side-effects or not.
+    """
+
+    @classmethod
+    @abc.abstractmethod
+    def has_effects(cls, op: Operation) -> bool:
+        raise NotImplementedError()
+
+
+def is_side_effect_free(op: Operation):
+    """
+    Boilerplate helper to check if a generic operation is side effect free for sure.
+    """
+    # If it doesn't say, safely assume it has side effects.
+    if not (trait := op.get_trait(MemoryEffect)):
+        return False
+    return not trait.has_effects(op)
+
```

Comment:
Maybe better on MemoryEffect?

Reply:
Well no, the main point of the helper is to say it might have side effect if the op does not implement the MemoryEffect interface (e.g., unregistered or just experimental new dialect!)

Reply:
I mean as a static function in the MemoryEffect class, as a namespacing thing